README.md
[![Build Status](https://travis-ci.org/n3okill/enfscopy.svg)](https://travis-ci.org/n3okill/enfscopy)
[![Build status](https://ci.appveyor.com/api/projects/status/wtrnfoek893y2x4p?svg=true)](https://ci.appveyor.com/project/n3okill/enfscopy)
[![Codacy Badge](https://api.codacy.com/project/badge/grade/478a5bf200e44bec8dbb6d6ed5e96dad)](https://www.codacy.com/app/n3okill/enfscopy)
[![Donate](https://www.paypalobjects.com/en_US/i/btn/btn_donate_SM.gif)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=64PYTCDH5UNZ6)

[![NPM](https://nodei.co/npm/enfscopy.png)](https://nodei.co/npm/enfscopy/)

enfscopy
=========
Module that add copy functionality to node fs module

**enfs** stands for [E]asy [N]ode [fs]

This module is intended to work as a sub-module of [enfs](https://www.npmjs.com/package/enfs)


Description
-----------
This module will add a method that allows the copy of multiple files
and folder in the file system, similar to cp -R

- This module will add following methods to node fs module:
  * copy
  * copySync
  
Usage
-----
`enfscopy`

```js
    var enfscopy = require("enfscopy");
```

Errors
------
All the methods follows the node culture.
- Async: Every async method returns an Error in the first callback parameter
- Sync: Every sync method throws an Error.


Additional Methods
------------------
- [copy](#copy)
- [copySync](#copysync)


### copy
  - **copy(srcPath, dstPath, [options], callback)**

> Asynchronously copy items in the file system


[options]:
  * fs (Object): an alternative fs module to use (default will be [enfspatch](https://www.npmjs.com/package/enfspatch))
  * limit (Integer): the limit number of files being copied at a moment (Default: fs module limit or 512)
  * overwrite (Boolean): if true will overwrite destination files if they exist before copy (Default: false)
  * preserveTimestamps (Boolean): if true will preserve the timestamps of copied items (Default: false)
  * stopOnError (Boolean): if true will stop copy execution at first error (Default: false)
  * dereference (Boolean): if true will dereference symlinks copying the items to where it points (default: false)
  * errors (Array or Stream): If array or stream, the errors that occur will be logged and returned


```js
    enfscopy.copy("/path/to/src/folder", "/path/to/destination", function(err, statistics){
        if(!err){
            console.log("Copied %d items with a total size of %d",statistics.items,statistics.size);
        }
    });
```


### copySync
  - **copySync(srcPath, dstPath, [options])**

> Synchronously copy items in the file system


[options]:
  * fs (Object): an alternative fs module to use (default will be [enfspatch](https://www.npmjs.com/package/enfspatch))
  * limit (Integer): the limit number of files being copied at a moment (Default: fs module limit or 512)
  * overwrite (Boolean): if true will overwrite destination files if they exist before copy (Default: false)
  * preserveTimestamps (Boolean): if true will preserve the timestamps of copied items (Default: false)
  * stopOnError (Boolean): if true will stop copy execution at first error (Default: false)
  * dereference (Boolean): if true will dereference symlinks copying the items to where it points (default: false)
  * errors (Array or Stream): If array or stream, the errors that occur will be logged and returned



```js
    var statistics = enfscopy.copySync("/path/to/src/folder", "/path/to/destination");
    console.log("Copied %d items with a total size of %d",statistics.items,statistics.size);
```

Credit
------
This functionality is based on [NCP](https://github.com/AvianFlu/ncp) with many improvements and a sync version.



License
-------

Creative Commons Attribution 4.0 International License

Copyright (c) 2016 Joao Parreira <joaofrparreira@gmail.com> [GitHub](https://github.com/n3okill)

This work is licensed under the Creative Commons Attribution 4.0 International License. 
To view a copy of this license, visit [CC-BY-4.0](http://creativecommons.org/licenses/by/4.0/).




CHANGELOG.md
1.0.0 / 2017-02-20
==================
  * code improvement
  * dependencies update
  * major version
  
0.1.0 / 2016-04-07
==================
  * es6 code update
  * dependencies update


0.0.1 / 2016-02-20
==================

  * first release


test/copySync.js
/* global afterEach, beforeEach, describe, it, after, before, process, __dirname */
/**
 * Created by n3okill on 22-12-2015.
 */


"use strict";


const nodePath = require("path");
const nodeOs = require("os");
const enFs = require("enfspatch");
const nodeCrypto = require("crypto");
const rimraf = require("rimraf");
const enfsmkdirp = require("enfsmkdirp");
const enFsCopy = require("../");
const copy = enFsCopy.copySync;
const cwd = process.cwd();

/*describe.skip("enFsCopySync", function () {
 let windowsTestLink = true;
 const tmpPath = nodePath.join(nodeOs.tmpdir(), "enfscopysync");
 const SIZE = 16 * 64 * 1024 + 7;
 const isWindows = /^win/.test(process.platform);

 before(function () {
 enfsmkdirp.mkdirpSync(tmpPath);
 process.chdir(tmpPath);
 if (isWindows) {
 enFs.writeFileSync(nodePath.join(tmpPath, "windowsTest", ""));
 try {
 enFs.symlinkSync(nodePath.join(tmpPath, "windowsTestLink"), nodePath.join(tmpPath, "windowsTest"), "file");
 } catch (err) {
 if (err.code === "EPERM") {
 windowsTestLink = false;
 }
 }
 }
 });
 /*afterEach(function() {
 rimraf.sync(tmpPath + nodePath.sep + "*");
 });
 after(function () {
 process.chdir(cwd);
 rimraf.sync(tmpPath);
 });
 describe("> when the source is a file", function () {
 it("should copy the file", function () {
 const src = nodePath.join(tmpPath, "TEST_enfscopy_src");
 const dst = nodePath.join(tmpPath, "TEST_enfscopy_dst");
 enFs.writeFileSync(src, nodeCrypto.randomBytes(SIZE));
 const srcMd5 = nodeCrypto.createHash("md5").update(enFs.readFileSync(src)).digest("hex");
 copy(src, dst);
 const dstMd5 = nodeCrypto.createHash("md5").update(enFs.readFileSync(dst)).digest("hex");
 srcMd5.should.be.equal(dstMd5);
 });
 it("should return an error if the source file does not exist", function (done) {
 const src = "this-file-does-not-exist.file";
 const dst = nodePath.join(tmpPath, "TEST_enfscopy_dst");
 try {
 copy(src, dst);
 } catch (err) {
 err.should.be.instanceOf(Error);
 done();
 }
 });
 it("should only copy files allowed by filter regex", function () {
 const src = nodePath.join(tmpPath, "file.bin");
 enFs.writeFileSync(src, "");
 const dst = nodePath.join(tmpPath, "dstfile.html");
 const filter = /.html$|.css$/i;
 copy(src, dst, filter);
 try {
 enFs.statSync(dst);
 } catch (err) {
 err.should.be.instanceOf(Error);
 }
 });
 it("should only copy files allowed by filter function", function () {
 const src = nodePath.join(tmpPath, "file.css");
 enFs.writeFileSync(src, "");
 const dst = nodePath.join(tmpPath, "dstFile.css");
 const filter = function (s) {
 return s.split(".").pop() !== "css";
 };
 copy(src, dst, filter);
 try {
 enFs.statSync(dst);
 } catch (err) {
 err.should.be.instanceOf(Error);
 }
 });
 it("accepts options object in place of filter", function () {
 const src = nodePath.join(tmpPath, "file.bin");
 enFs.writeFileSync(src, "");
 const dst = nodePath.join(tmpPath, "dstFile.bin");
 copy(src, dst, {filter: /.html$|.css$/i});
 try {
 enFs.statSync(dst);
 } catch (err) {
 err.should.be.instanceOf(Error);
 }
 });
 it("should maintain file mode", function () {
 const src = nodePath.join(tmpPath, "TEST_enfscopy_src");
 const dst = nodePath.join(tmpPath, "TEST_enfscopy_dst");
 enFs.writeFileSync(src, nodeCrypto.randomBytes(SIZE));
 enFs.chmodSync(src, parseInt("0750", 8));
 copy(src, dst);
 enFs.statSync(src).mode.should.be.equal(enFs.statSync(dst).mode);
 });

 describe("> when the destination dir does not exist", function () {
 it("should create the destination directory and copy the file", function () {
 const src = nodePath.join(tmpPath, "file.txt");
 const data = "did it copy?\n";
 enFs.writeFileSync(src, data, "utf8");
 const dst = nodePath.join(tmpPath, "this", "path", "does", "not", "exist", "copied.txt");
 copy(src, dst);
 enFs.readFileSync(dst, "utf8").should.be.equal(data);
 });
 });
 });
 describe("> when the source is a directory", function () {
 describe("> when the source directory does not exist", function () {
 it("should return an error", function (done) {
 const src = nodePath.join(tmpPath, "this_dir_dos_not_exist");
 const dst = nodePath.join(tmpPath, "this_dir_really_does_not_matter");
 try {
 copy(src, dst);
 } catch (err) {
 err.should.be.instanceOf(Error);
 done();
 }
 });
 });
 it("should copy the directory", function () {
 const FILES = 2;
 const src = nodePath.join(tmpPath, "src");
 const dst = nodePath.join(tmpPath, "dst");
 enFs.mkdirSync(src);
 for (let i = 0; i < FILES; i++) {
 enFs.writeFileSync(nodePath.join(src, i.toString()), nodeCrypto.randomBytes(SIZE));
 }
 const subdir = nodePath.join(src, "subdir");
 enFs.mkdirSync(subdir);
 for (let i = 0; i < FILES; i++) {
 enFs.writeFileSync(nodePath.join(subdir, i.toString()), nodeCrypto.randomBytes(SIZE));
 }
 copy(src, dst);
 const statDst = enFs.statSync(dst);
 let statFile;
 statDst.isDirectory().should.be.equal(true);
 for (let i = 0; i < FILES; i++) {
 statFile = enFs.statSync(nodePath.join(dst, i.toString()));
 statFile.isFile().should.be.equal(true);
 }
 const dstSubDir = nodePath.join(dst, "subdir");
 const statDstSubDir = enFs.statSync(dstSubDir);
 statDstSubDir.isDirectory().should.be.equal(true);
 for (let i = 0; i < FILES; i++) {
 statFile = enFs.statSync(nodePath.join(dstSubDir, i.toString()));
 statFile.isFile().should.be.equal(true);
 }
 });
 describe("> when the destination directory does not exist", function () {
 it("should create the destination directory and copy the file", function () {
 const src = nodePath.join(tmpPath, "data");
 enFs.mkdirSync(src);
 const f1 = "file1";
 const f2 = "file2";
 enFs.writeFileSync(nodePath.join(src, "f1.txt"), f1);
 enFs.writeFileSync(nodePath.join(src, "f2.txt"), f2);
 const dst = nodePath.join(tmpPath, "this", "path", "does", "not", "exist");
 copy(src, dst);
 const c1 = enFs.readFileSync(nodePath.join(dst, "f1.txt"), "utf8");
 const c2 = enFs.readFileSync(nodePath.join(dst, "f2.txt"), "utf8");
 c1.should.be.equal(f1);
 c2.should.be.equal(f2);
 });
 });
 describe("> when src directory does not exist", function () {
 it("should return an error", function (done) {
 try {
 copy("/path/does/not/exist", "/to/any/place");
 } catch (err) {
 err.should.be.instanceOf(Error);
 done();
 }
 });
 });
 describe("> when overwrite is true and dst is readonly", function () {
 it("should copy the file and not throw an error", function () {
 const src = nodePath.join(tmpPath, "TEST_enfscopy_src");
 const dst = nodePath.join(tmpPath, "TEST_enfscopy_dst");
 enFs.writeFileSync(src, "datasrc");
 enFs.writeFileSync(dst, "data");
 enFs.chmodSync(dst, parseInt("0444", 8));
 copy(src, dst, {overwrite: true});
 const contents = enFs.readFileSync(dst, "utf8");
 contents.should.containEql("datasrc");
 });
 });
 });
 describe("> when the src and dst are identical", function () {
 let file, fileData;
 before(function () {
 file = nodePath.join(tmpPath, "identicalFile");
 });
 beforeEach(function () {
 fileData = "some data";
 enFs.writeFileSync(file, fileData);
 });
 describe("> when the src and dst are the same file", function () {
 it("should not copy and not throw an error", function () {
 copy(file, file);
 enFs.readFileSync(file, "utf8").should.be.equal(fileData);
 });
 });
 describe("> when the src is symlink and points to dst", function () {
 it("should not copy and keep the symlink", function () {
 const src = nodePath.join(tmpPath, "testLink");
 const dst = file;
 if (isWindows && !windowsTestLink) {
 return;
 }
 enFs.symlinkSync(file, src);
 copy(src, dst);
 enFs.readFileSync(dst, "utf8").should.be.equal(fileData);
 enFs.readlinkSync(src).should.be.equal(dst);
 });
 });
 describe("> when dst is symlink and points to src", function () {
 it("should not copy and keep the symlink", function () {
 const src = file;
 const dst = nodePath.join(tmpPath, "testLink");
 if (isWindows && !windowsTestLink) {
 return;
 }
 enFs.symlinkSync(file, dst);
 copy(src, dst);
 enFs.readFileSync(dst, "utf8").should.be.equal(fileData);
 enFs.readlinkSync(dst).should.be.equal(src);
 });
 });
 });
 describe("> when using dereference", function () {
 let src, file, dir, dst;
 beforeEach(function () {
 src = nodePath.join(tmpPath, "src");
 file = nodePath.join(src, "file");
 const fileLink = nodePath.join(src, "fileLink");
 dir = nodePath.join(src, "dir");
 const dirFile = nodePath.join(dir, "dirFile");
 const dirLink = nodePath.join(src, "dirFileLink");
 dst = nodePath.join(tmpPath, "dst");
 if (isWindows && !windowsTestLink) {
 return;
 }
 enFs.mkdirSync(src);
 enFs.writeFileSync(file, "contents");
 enFs.symlinkSync(file, fileLink, "file");
 enFs.mkdirSync(dir);
 enFs.writeFileSync(dirFile, "contents");
 enFs.symlinkSync(dir, dirLink, "dir");
 });
 it("copies symlinks by default", function () {
 if (isWindows && !windowsTestLink) {
 return;
 }
 copy(src, dst);
 enFs.readlinkSync(nodePath.join(dst, "fileLink")).should.be.equal(file);
 enFs.readlinkSync(nodePath.join(dst, "dirFileLink")).should.be.equal(dir);
 });

 it("copies file contents when dereference=true", function () {
 if (isWindows && !windowsTestLink) {
 return;
 }
 copy(src, dst, {dereference: true});
 const fileSymLink = nodePath.join(dst, "fileLink");
 const dirSymLink = nodePath.join(dst, "dirFileLink");
 enFs.statSync(fileSymLink).isFile().should.be.equal(true);
 enFs.readFileSync(fileSymLink, "utf8").should.be.equal("contents");
 enFs.statSync(dirSymLink).isDirectory().should.be.equal(true);
 enFs.readdirSync(dirSymLink).should.be.eql(["dirFile"]);
 });
 });
 });*/


describe("enFsCopySync", function () {
    let windowsTestLink = true;
    const tmpPath = nodePath.join(nodeOs.tmpdir(), "enfscopysync");
    const SIZE = 16 * 64 * 1024 + 7;
    const isWindows = /^win/.test(process.platform);
    let FILES = 2;

    before(function () {
        enfsmkdirp.mkdirpSync(tmpPath);
        process.chdir(tmpPath);
        if (isWindows) {
            enFs.writeFileSync(nodePath.join(tmpPath, "windowsTest", ""));
            try {
                enFs.symlinkSync(nodePath.join(tmpPath, "windowsTestLink"), nodePath.join(tmpPath, "windowsTest"), "file");
            } catch (err) {
                if (err.code === "EPERM") {
                    console.warn("Not running tests with symlink in windows, don't have permissions, \nif you wan't to test this try running as administrator.");
                    windowsTestLink = false;
                }
            }
        }
        enFs.writeFileSync(nodePath.join(tmpPath, "TEST_enfscopy_src"), nodeCrypto.randomBytes(SIZE));
        enFs.writeFileSync(nodePath.join(tmpPath, "file.bin"), "");
        enFs.writeFileSync(nodePath.join(tmpPath, "file.css"), "");
        enFs.writeFileSync(nodePath.join(tmpPath, "file1.bin"), "");
        enFs.writeFileSync(nodePath.join(tmpPath, "file.txt"), "did it copy?\n", "utf8");
        enFs.mkdirSync(nodePath.join(tmpPath, "srca"));
        for (let i = 0; i < FILES; i++) {
            enFs.writeFileSync(nodePath.join(tmpPath, "srca", i.toString()), nodeCrypto.randomBytes(SIZE));
        }
        enFs.mkdirSync(nodePath.join(tmpPath, "srca", "subdir"));
        for (let i = 0; i < FILES; i++) {
            enFs.writeFileSync(nodePath.join(tmpPath, "srca", "subdir", i.toString()), nodeCrypto.randomBytes(SIZE));
        }
        enFs.mkdirSync(nodePath.join(tmpPath, "data"));
        enFs.writeFileSync(nodePath.join(tmpPath, "data", "f1.txt"), "file1");
        enFs.writeFileSync(nodePath.join(tmpPath, "data", "f2.txt"), "file2");
        enFs.writeFileSync(nodePath.join(tmpPath, "identicalFile"), "some data");
        enFs.writeFileSync(nodePath.join(tmpPath, "identicalFile1"), "some data");
        if (windowsTestLink) {
            enFs.symlinkSync(nodePath.join(tmpPath, "identicalFile"), nodePath.join(tmpPath, "testLink"), "file");
            enFs.symlinkSync(nodePath.join(tmpPath, "identicalFile1"), nodePath.join(tmpPath, "testLink1"), "file");
        }
    });
    after(function () {
        process.chdir(cwd);
        rimraf.sync(tmpPath);
    });
    describe("> when the source is a file", function () {
        it("should copy the file", function () {
            const src = nodePath.join(tmpPath, "TEST_enfscopy_src");
            const dst = nodePath.join(tmpPath, "TEST_enfscopy_dst");
            const srcMd5 = nodeCrypto.createHash("md5").update(enFs.readFileSync(src)).digest("hex");
            copy(src, dst);
            const dstMd5 = nodeCrypto.createHash("md5").update(enFs.readFileSync(dst)).digest("hex");
            srcMd5.should.be.equal(dstMd5);
        });
        it("should return an error if the source file does not exist", function () {
            const src = "this-file-does-not-exist.file";
            const dst = nodePath.join(tmpPath, "TEST_enfscopy_dst");
            (function () {
                copy(src, dst);
            }).should.throw(Error);
        });
        it("should only copy files allowed by filter regex", function () {
            const src = nodePath.join(tmpPath, "file.bin");
            const dst = nodePath.join(tmpPath, "dstfile.html");
            const filter = /.html$|.css$/i;
            copy(src, dst, filter);
            (function () {
                enFs.statSync(dst);
            }).should.throw(Error);
        });
        it("should only copy files allowed by filter function", function () {
            const src = nodePath.join(tmpPath, "file.css");
            const dst = nodePath.join(tmpPath, "dstFile.css");
            const filter = function (path) {
                return path.split(".").pop() !== "css";
            };
            copy(src, dst, filter);
            (function () {
                enFs.statSync(dst);
            }).should.throw(Error);
        });
        it("accepts options object in place of filter", function () {
            const src = nodePath.join(tmpPath, "file1.bin");
            const dst = nodePath.join(tmpPath, "dstFile.bin");
            copy(src, dst, {filter: /.html$|.css$/i});
            (function () {
                enFs.statSync(dst);
            }).should.throw(Error);
        });
        describe("> when the destination dir does not exist", function () {
            it("should create the destination directory and copy the file", function () {
                const src = nodePath.join(tmpPath, "file.txt");
                const dst = nodePath.join(tmpPath, "this", "path", "does", "not", "exist", "copied.txt");
                copy(src, dst);
                enFs.readFileSync(dst, "utf8").should.be.equal("did it copy?\n");
            });
        });
    });
    describe("> when the source is a directory", function () {
        describe("> when the source directory does not exist", function () {
            it("should return an error", function () {
                const src = nodePath.join(tmpPath, "this_dir_dos_not_exist");
                const dst = nodePath.join(tmpPath, "this_dir_really_does_not_matter");
                (function () {
                    copy(src, dst);
                }).should.throw(Error);
            });
        });
        it("should copy the directory", function () {
            let FILES = 2;
            let copiedSize = 0;
            const src = nodePath.join(tmpPath, "srca");
            const dst = nodePath.join(tmpPath, "dsta");
            let stats = copy(src, dst);

            let statFile;
            const statDst = enFs.statSync(dst);
            statDst.isDirectory().should.be.equal(true);
            for (let i = 0; i < FILES; i++) {
                statFile = enFs.statSync(nodePath.join(dst, i.toString()));
                statFile.isFile().should.be.equal(true);
                copiedSize += statFile.size;
            }
            const dstSubDir = nodePath.join(dst, "subdir");
            const statDstSubDir = enFs.statSync(dstSubDir);
            statDstSubDir.isDirectory().should.be.equal(true);
            for (let i = 0; i < FILES; i++) {
                statFile = enFs.statSync(nodePath.join(dstSubDir, i.toString()));
                statFile.isFile().should.be.equal(true);
                copiedSize += statFile.size;
            }
            copiedSize.should.be.equal(stats.copied.size);
        });
        describe("> when the destination directory does not exist", function () {
            it("should create the destination directory and copy the file", function () {
                const src = nodePath.join(tmpPath, "data");
                const dst = nodePath.join(tmpPath, "this", "path", "does", "not", "exist");
                copy(src, dst);
                enFs.readFileSync(nodePath.join(dst, "f1.txt"), "utf8").should.be.equal("file1");
                enFs.readFileSync(nodePath.join(dst, "f2.txt"), "utf8").should.be.equal("file2");
            });
        });
        describe("> when src directory does not exist", function () {
            it("should return an error", function () {
                (function () {
                    copy("/path/does/not/exist", "/to/any/place");
                }).should.throw(Error);
            });
        });
    });
    describe("> when the src and dst are identical", function () {
        const file = nodePath.join(tmpPath, "identicalFile");
        const file1 = nodePath.join(tmpPath, "identicalFile1");
        const fileData = "some data";
        describe("> when the src and dst are the same file", function () {
            it("should not copy and not throw an error", function () {
                copy(file, file);
                enFs.readFileSync(file, "utf8").should.be.equal(fileData);
            });
        });
        describe("> when the src is symlink and points to dst", function () {
            it("should not copy and keep the symlink", function () {
                const src = nodePath.join(tmpPath, "testLink");
                const dst = file;
                if (isWindows && !windowsTestLink) {
                    return;
                }
                copy(src, dst);
                enFs.readFileSync(dst, "utf8").should.be.equal(fileData);
                enFs.readlinkSync(src).should.be.equal(dst);
            });
        });
        describe("> when dst is symlink and points to src", function () {
            it("should not copy and keep the symlink", function () {
                const src = file1;
                const dst = nodePath.join(tmpPath, "testLink1");
                if (isWindows && !windowsTestLink) {
                    return;
                }

                copy(src, dst);
                enFs.readFileSync(dst, "utf8").should.be.equal(fileData);
                enFs.readlinkSync(dst).should.be.equal(src);
            });
        });
    });
    describe("> when using dereference", function () {
        before(function () {
            if (windowsTestLink) {
                enfsmkdirp.mkdirpSync(nodePath.join(tmpPath, "src", "default"));
                enFs.writeFileSync(nodePath.join(tmpPath, "src", "default", "file"), "contents");
                enFs.symlinkSync(nodePath.join(tmpPath, "src", "default", "file"), nodePath.join(tmpPath, "src", "default", "fileLink"), "file");
                enFs.mkdirSync(nodePath.join(tmpPath, "src", "default", "dir"));
                enFs.writeFileSync(nodePath.join(tmpPath, "src", "default", "dirFile"), "contents");
                enFs.symlinkSync(nodePath.join(tmpPath, "src", "default", "dir"), nodePath.join(tmpPath, "src", "default", "dirFileLink"), "dir");
                enfsmkdirp.mkdirpSync(nodePath.join(tmpPath, "src", "deref"));
                enFs.writeFileSync(nodePath.join(tmpPath, "src", "deref", "file"), "contents");
                enFs.symlinkSync(nodePath.join(tmpPath, "src", "deref", "file"), nodePath.join(tmpPath, "src", "deref", "fileLink"), "file");
                enFs.mkdirSync(nodePath.join(tmpPath, "src", "deref", "dir"));
                enFs.writeFileSync(nodePath.join(tmpPath, "src", "deref", "dir", "dirFile"), "contents");
                enFs.symlinkSync(nodePath.join(tmpPath, "src", "deref", "dir"), nodePath.join(tmpPath, "src", "deref", "dirFileLink"), "dir");
            }
        });
        it("copies symlinks by default", function () {
            const src = nodePath.join(tmpPath, "src", "default");
            const file = nodePath.join(src, "file");
            const dir = nodePath.join(src, "dir");
            const dst = nodePath.join(tmpPath, "dst", "default");
            if (isWindows && !windowsTestLink) {
                return;
            }
            copy(src, dst);
            enFs.readlinkSync(nodePath.join(dst, "fileLink")).should.be.equal(file);
            enFs.readlinkSync(nodePath.join(dst, "dirFileLink")).should.be.equal(dir);
        });

        it("copies file contents when dereference=true", function () {
            const src = nodePath.join(tmpPath, "src", "deref");
            const dst = nodePath.join(tmpPath, "dst", "deref");
            if (isWindows && !windowsTestLink) {
                return;
            }
            copy(src, dst, {dereference: true});
            const fileSymLink = nodePath.join(dst, "fileLink");
            const dirSymLink = nodePath.join(dst, "dirFileLink");
            enFs.statSync(fileSymLink).isFile().should.be.equal(true);
            enFs.readFileSync(fileSymLink, "utf8").should.be.equal("contents");
            enFs.statSync(dirSymLink).isDirectory().should.be.equal(true);
            enFs.readdirSync(dirSymLink).should.be.eql(["dirFile"]);
        });
    });
});

lib/copySync.js
/**
 * @project enfscopy
 * @filename copySync.js
 * @description sync method to copy items in the file system
 * @author Joao Parreira <joaofrparreira@gmail.com>
 * @copyright Copyright(c) 2016 Joao Parreira <joaofrparreira@gmail.com>
 * @licence Creative Commons Attribution 4.0 International License
 * @createdAt Created at 18-02-2016.
 * @version 0.0.2
 */


"use strict";


const nodePath = require("path");
const enFs = require("enfspatch");
const enfsmkdirp = require("enfsmkdirp").mkdirpSync;
const enfsfind = require("enfsfind").findSync;


const BUF_LENGTH = 64 * 1024;


const kindOf = (arg) => arg === null ? "null" : typeof arg === "undefined" ? "undefined" : /^\[object (.*)\]$/.exec(Object.prototype.toString.call(arg))[1].toLowerCase();
const isFunction = (arg) => "function" === kindOf(arg);
const isRegExp = (arg) => "regexp" === kindOf(arg);


class CopyFiles {
    constructor(src, dst, options) {
        this.options = options;
        this.basePath = process.cwd();
        this.src = nodePath.resolve(this.basePath, src);
        this.dst = nodePath.resolve(this.basePath, dst);
        this.stat = options.dereference ? options.fs.statSync : options.fs.lstatSync;
        this.mkdirp = options.fs.mkdirpSync || enfsmkdirp;
        this.BUF_LENGTH = "BUF_LENGTH" in options ? options.BUF_LENGTH : BUF_LENGTH;
        this._buff = new Buffer(this.BUF_LENGTH);

        this.itemsToCopy = [];
        this.statistics = {
            copied: {
                files: 0,
                directories: 0,
                links: 0,
                size: 0
            },
            overwrited: 0,
            skipped: 0,
            size: 0,
            errors: 0,
            items: 0
        };
        this.hasErrors = false;
        this.errors = null;
        if (!options.stopOnError) {
            if (options.errors) {
                this.errors = options.fs.createWriteStream(options.errors);
            } else {
                this.errors = [];
            }
        }
    }


    start() {
        if (this.src === this.dst) {
            return this.statistics;
        }
        if ((this.dst + nodePath.sep).indexOf(this.src + nodePath.sep) === 0) {
            let e = new Error("Error trying to copy. 'destination' is a sub-directory of 'source'");
            e.code = "ESELF";
            throw e;
        }

        this.stat(this.src);
        this.mkdirp(nodePath.dirname(this.dst), {fs: this.options.fs});
        return this.loadItems();
    }


    loadItems() {
        const items = enfsfind(this.src, {
            fs: this.options.fs,
            filter: this.options.filter,
            dereference: this.options.dereference
        });
        if (items.length === 0) {
            return this.statistics;
        }
        this.itemsToCopy = items;
        return this.loadStatistics(items);
    }


    loadStatistics(items) {
        this.statistics.items = items.length;
        this.statistics.size = items.map((item) => item.stat.isFile() ? item.stat.size : 0).reduce((acc, val) => acc + val, 0);
        return this.startCopy();
    }

    startCopy() {
        let items = this.itemsToCopy;
        do {
            const item = items.shift();
            if (item.stat.isDirectory()) {
                this.makeDir(item);
            } else if (item.stat.isFile() || item.stat.isBlockDevice() || item.stat.isCharacterDevice()) {
                this.onFile(item);
            } else if (item.stat.isSymbolicLink()) {
                this.onLink(item);
            }
        } while (items.length);
        return this.statistics;
    }


    makeDir(item) {
        const target = item.path.replace(this.src, this.dst);
        this.mkdirp(target, {fs: this.options.fs, mode: item.stat.mode});
        this.statistics.copied.directories++;
        this.options.fs.chmodSync(target, item.stat.mode);
    }


    isWritable(target) {
        try {
            this.stat(target);
        } catch (err) {
            if (err.code === "ENOENT") {
                return true;
            }
        }
        return false;
    }


    onFile(item) {
        const target = item.path.replace(this.src, this.dst);
        if (this.isWritable(target)) {
            this.copyFile(item, target);
        } else {
            if (this.options.overwrite && item.path !== target) {
                this.rmFile(target);
                this.copyFile(item, target);
            }
        }
    }


    rmFile(target) {
        this.options.fs.unlinkSync(target);
        this.statistics.overwrited++;
    }

    copyFile(item, target) {
        let bytesRead, pos;
        const fdr = this.options.fs.openSync(item.path, "r");
        const fdw = this.options.fs.openSync(target, "w", item.stat.mode);
        bytesRead = 1;
        pos = 0;
        while (bytesRead > 0) {
            bytesRead = this.options.fs.readSync(fdr, this._buff, 0, this.BUF_LENGTH, pos);
            this.options.fs.writeSync(fdw, this._buff, 0, bytesRead);
            pos += bytesRead;
        }
        if (this.options.preserveTimestamps) {
            this.options.fs.futimesSync(fdw, item.stat.atime, item.stat.mtime);
        }
        this.options.fs.closeSync(fdr);
        this.options.fs.closeSync(fdw);
        this.statistics.copied.files++;
        this.statistics.copied.size += item.stat.size;
    }


    onLink(item) {
        const target = item.path.replace(this.src, this.dst);
        const resolvedPath = this.options.fs.readlinkSync(item.path);
        this.checkLink(resolvedPath, target, item);
    }


    checkLink(resolvedPath, target, item) {
        if (this.options.dereference) {
            resolvedPath = nodePath.resolve(this.basePath, resolvedPath);
        }
        if (resolvedPath === target) {
            this.statistics.copied.link++;
            return;
        }
        if (this.isWritable(target)) {
            this.makeLink(resolvedPath, target, item);
        } else {
            let targetDst;
            try {
                targetDst = this.options.fs.readlinkSync(target);
            } catch (err) {
                if (err.code === "EINVAL") {
                    this.statistics.copied.links++;
                    return;
                }
                throw err;
            }
            if (this.options.dereference) {
                targetDst = nodePath.resolve(this.basePath, targetDst);
            }
            if (targetDst === resolvedPath) {
                this.statistics.copied.links++;
                return;
            }
            this.rmFile(target);
            this.makeLink(resolvedPath, target, item);
        }
    }

    makeLink(resolvedPath, target) {
        this.options.fs.symlinkSync(resolvedPath, target);
        this.statistics.copied.links++;
    }
}


/**
 * Copy items in the file system async
 * @param {string} src - the path to be copied
 * @param {string} dst - the destination path for the items
 * @param {object} opt - various options for copy module
 *              {object} fs - the fs module to be used
 *              {limit} limit - the maximum number of items being copied at same time
 *              {bool} overwrite - if true will overwrite destination items if existent
 *              {bool} preserveTimestamps - if true will preserve the timestamps of copied items
 *              {bool} stopOnError - if true will stop copying at first error
 *              {Array|WriteStream} errors - can be a array or WriteStream to where error's will be saved
 *              {bool} dereference - defines the type of stat that will be used in files
 *              {function|RegExp} filter - function or RegExp used to narrow results
 *              this function receives an object of type {path: itemPath, stat: itemStatObject}
 * @return {Error|Object} Error | Copied items statistics
 */
function copy(src, dst, opt) {
    if (isFunction(opt) || isRegExp(opt)) {
        opt = {filter: opt};
    }
    let options = opt || {};
    options.fs = options.fs || enFs;
    options.limit = options.limit || enFs.limit || 512;
    options.overwrite = options.overwrite === true;
    options.preserveTimestamps = options.preserveTimestamps === true;
    options.stopOnError = options.stopOnError !== false;
    options.dereference = options.dereference === true;
    options.errors = options.errors || null;
    return (new CopyFiles(src, dst, options)).start();
}


module.exports = copy;

